### DISCUSSION TITLE
scope value type

### DISCUSSION THREAD

**Discussion by acivitillo** [idom-team/idom#717] [Question]
We are building components on top of IDOM. One of such components is `BatchTable`, this is a Table component on which users can perform batch actions by clicking on checkboxes (see below).

I am wondering on what would be the best practice to keep state here? I am considering using a string and then send an event on each checkbox check/uncheck to update such state. Does that make sense?

I am asking because `use_state()` doesn't seem to support `dict`, so I cannot do `value, set_value = use_state({})` for instance (probably for good reasons).

So the structure I had in mind was a string of ids with some delimiter to reflect the checked boxes on the `BatchTable`.

![image](https://user-images.githubusercontent.com/54931660/160272162-38555d1a-0c53-4f5c-bd66-27c5ab4061b1.png)

**Comment by rmorshea** [comment]
You're free to use dicts with `use_state`, you just can't mutate the value and trigger a re-render. We discuss some good ways of working with dicts and Python's other built-in, but mutable data types [in the docs](https://idom-docs.herokuapp.com/docs/guides/adding-interactivity/dangers-of-mutability/index.html#updating-dictionary-items). If you're running into troubles with mutable data types though, I'd recommend looking into [`pyrsistent`](https://pypi.org/project/pyrsistent). It's a framework for working with immutable data types.

**Reply by Archmonger** [reply]
To avoid adding additional dependencies, I personally have been using the built-in `dataclass` with `frozen=true`.

https://docs.python.org/3/library/dataclasses.html#frozen-instances

**Reply by Archmonger** [reply]
On this note, we probably should log a warning for mutable types when `IDOM_DEBUG_MODE=True`.

**Reply by rmorshea** [reply]
@acivitillo, I'd actually imagine you could get away with using a `set` instead of `dict` for what you're trying to do:

```python
def use_checked():
    checked, set_checked = use_state(set())

    def is_checked(item):
        return item in checked
    
    def toggle_check(item):
        if item in checked:
            set_data(data - {item})
        else:
            set_data(data | {item})
    
    return is_checked, toggle_check


@component
def BatchTable():
    is_checked, toggle_check = use_checked()
    ...
```

**Reply by rmorshea** [reply]
@Archmonger, I'm not sure how we'd produce that warning. It's not wrong to use mutable data types so long as you know what you're doing. All of Javascript's built-in data types are mutable and React does not produce a similar warning.

**Reply by rmorshea** [reply]
@acivitillo it'd be great to re-sync at some point. I'd like to hear about what has/hasn't gone well so far.

**Comment by acivitillo** [comment]
yes I would love to, how you suggest we do that?

>

**Reply by rmorshea** [reply]
Feel free to book another [office hour slot](https://calendly.com/rmorshea/60min?month=2022-03). If that doesn't seem like it will work let me know another time you'd prefer.
